src/shutdown.rs

```diff
@@ -0,0 +1,253 @@
+use crate::runtime;
+use futures01::{future, stream::Stream, Async, Future};
+use std::collections::HashMap;
+use std::sync::Arc;
+use std::time::{Duration, Instant};
+use stream_cancel::{Trigger, Tripwire};
+use tokio::timer;
+
+/// When this struct goes out of scope and its internal refcount goes to 0 it is a signal that its
+/// corresponding Source has completed executing and may be cleaned up.  It is the responsibility
+/// of each Source to ensure that at least one copy of this handle remains alive for the entire
+/// lifetime of the Source.
+#[derive(Clone)]
+pub struct ShutdownCompleteHandle {
+    _shutdown_complete: Arc<Trigger>,
```

Comment:
Okay, so just to be clear here we never actually use `Trigger::cancel` but instead use its drop impl + Arc's ref count to achieve what we want? Seems like a good idea!

Reply:
Thanks!  I'm pretty satisfied with how this part turned out.  Although you deserve most of the credit for the idea!